site/content/blog/2021-06-01-migrating-operator-custom-resource.md

```diff
@@ -0,0 +1,142 @@
+
+## Objectives
+
+As part of [Mattermost Cloud offering](https://mattermost.com/mattermost-cloud/), we have thousands of Mattermost instances running in our Kubernetes clusters and each of them is represented as a CR managed by the Mattermost Operator.
+
+Mattermost Operator is also one of [the recommended ways to install Mattermost](https://docs.mattermost.com/guides/administrator.html#installing-mattermost) for self-managed customers.
+
+Those facts put some constraints on our migration story and we boiled them down to the following:
+- No downtime of Mattermost instance during the migration.
+- As little manual intervention as possible.
+- Ability to run the migration for several installations at the same time.
+
+## Changing the Custom Resource
+
+After using Mattermost Operator for quite some time, we've learned some lessons and decided to simplify some things.
+
+However, removing some features and adding new ones was not the biggest change in the migration process.
+
+Our initial CRs representing Mattermost installations were called `ClusterInstallation`, which made sense from the domain perspective of Mattermost Cloud but didn't really follow good practices of the Kubernetes ecosystem.
+For a `v1beta1` specification we decided to change the whole CR to simply `Mattermost`.
+
+## Migration challenges
+
+### Changing resource Kind
+
+The usual way of migrating CRs to the new version is by using [Conversion Webhooks](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion) which give the ability to convert the CR specification "on the fly" by creating a webhook used by the Kubernetes API server. This mechanism allows users to operate on several versions of resource while only one version is stored in the `etcd` database.
+
+Conversion webhooks however are designed to handle the migration from one version to another but the Custom Resource name and group can't be changed during this process as those are the identifying properties of the Custom Resource Definition (CRD).
+
+Unfortunately, conversion webhooks did not apply to our case as in addition to the change of version, we also changed both resource Group and Kind, effectively creating a whole new CR - `Mattermost`.
+
+#### Migrating to the new resource
+
+When the `ClusterInstallation` CR is created, the Mattermost Operator spins up a new instance of the application. This includes Kubernetes Deployments which are running the actual Mattermost application. When the CR is deleted all the resources are deleted together with it thanks to [Owner References](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents).
+
+This prevents us from simply running the script that would delete `ClusterInstallation` and create `Mattermost` in its place as it would cause the downtime of the Mattermost application (unless we chose to orphan all resources).
```

Comment:
```suggestion
This prevents us from simply running the script that would delete `ClusterInstallation` and create `Mattermost` in its place as it would cause the downtime of the Mattermost application (unless we choose to orphan all resources).
```